### PR TITLE
Use NuGet 3.x for compatibility and for Get Nearest

### PR DESCRIPTION
**Pull request by joelverhagen**
### Done
- Add support for TFM compatibilities that are in 3.x but not in 2.x (addresses https://github.com/NuGet/Home/issues/2720).
- Update TFM precedence to match 3.x. This is the Get Nearest algorithm to select which TFM to use given a single project TFM and one or more package TFMs. The old precedence algorithm (confusing scoring thing) has been removed.
- Add `NuGet.Frameworks` as a dependency of `NuGet.Core`. Note that the 3.x code that depends on `NuGet.Core` should take its own self-consistent version of `NuGet.Frameworks.`.
- Clean up some bad static stuff which made testing messing.
### Questions
- What release of 3.x should 2.12 depend on?
- What source will have NuGet.Frameworks? A change was needed to NuGet.Frameworks so that `net40-client` was a supported TFM.
- Is the additional `NuGet.Frameworks.dll` in the VSIX okay? Similarly, is the `NuGet.Frameworks` dependency in the NuGet.Core package okay? If not, we can ILMerge `NuGet.Frameworks.dll` into `NuGet.Core.dll`.

@emgarten @rohit21agrawal @yishaigalatzer @rrelyea @alpaix

### CODE REVIEW

**Comment by emgarten** [comment]
We'll need to make sure that the NuGet.Frameworks project is _nearer_ when integrating this into nuget.client. For this reason I don't think we can ILMerge these, it wouldn't be possible to override then.

This change looks good overall to me, it's amazing it came out as clean as it did :smile: 

:shipit:

**Comment by joelverhagen** [comment]
I've updated this to pull in NuGet.Frameworks as a submodule. This is enabled by https://github.com/NuGet/NuGet.Client/pull/595.

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 1]
```diff
@@ -80,7 +94,7 @@ internal static bool HasCompatibleProfileWith(NetPortableProfile packageFramewor
             return false;
         }
 
-        private static NetPortableProfileCollection BuildPortableProfileCollection()
+        public static NetPortableProfileCollection BuildPortableProfileCollection()
```
Comment (rohit21agrawal) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: suggestion : if this isn't being called from another assembly, can we make it internal? (similarly at other places too)

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 2]
```diff
@@ -285,7 +299,9 @@ public CompiledNetPortableProfileCollection(NetPortableProfileCollection profile
 
             private static IDictionary<string, NetPortableProfile> CreatePortableProfilesByCustomProfileString(NetPortableProfileCollection profileCollection)
             {
-                return profileCollection.ToDictionary(x => x.CustomProfileString);
+                return profileCollection
+                    .ToLookup(x => x.CustomProfileString)
+                    .ToDictionary(g => g.Key, g => g.Last());
```
Comment (rohit21agrawal) [inline src/Core/NETPortable/NetPortableProfileTable.cs:306]: nit: can you update the comment, and explain the choice of key and value.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 1 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: I fixed this one, but I'd rather make exhaustively flip public to internal.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 2 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs:306]: Fixed.

**Pull request merged**